### PR TITLE
Support for Interface and Unions in Nadel

### PR DESCRIPTION
**Pull request by bbakerman**
This does not allow for hydration of interfaces and unions yet.  I plan to get that in either on this PR or another one based on it

I also have not tested type renames (yet) but plan to as well.

### CODE REVIEW

File: src/main/java/graphql/nadel/Nadel.java [hunk 1]
```diff
@@ -50,7 +53,7 @@
 
     private final Logger log = LoggerFactory.getLogger(Nadel.class);
     private final StitchingDsl stitchingDsl;
-    private final ServiceDataFactory serviceDataFactory;
+    private final ServiceExecutionFactory serviceExecutionFactory;
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/Nadel.java]: I had to change the ServiceDataFactory signature - I took the opportunity to rename interface.  This is a breaking change either way

File: src/main/java/graphql/nadel/ServiceExecutionFactory.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    /**
+     * Called to return underlying type definitions for a schema
+     *
+     * @param serviceName the name of the service
+     *
+     * @return a {@link graphql.schema.idl.TypeDefinitionRegistry} of all the underlying schema types
+     */
+    TypeDefinitionRegistry getUnderlyingTypeDefinitions(String serviceName);
+}
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/ServiceExecutionFactory.java]: The key change is that we now ask services to return a parsed set of types and not a implemented schema.

We will implement it for them (as a noop).

File: src/main/java/graphql/nadel/engine/OverallQueryTransformer.java [hunk 3]
```diff
@@ -308,17 +311,24 @@ private TypeTransformation typeTransformationForFragment(ExecutionContext execut
...
+            fieldTransformation = new FieldRenameTransformation(definition.getFieldMappingDefinition());
         } else if (definition.getInnerServiceHydration() != null) {
-            return new HydrationTransformation(definition.getInnerServiceHydration());
+            fieldTransformation = new HydrationTransformation(definition.getInnerServiceHydration());
+            usedHydration = true;
+        }
+
+        if (UnderscoreTypeNameTransformation.isInterfaceOrUnionField(environment) && !usedHydration) {
+            fieldTransformation = new ComposedFieldTransformation(fieldTransformation, new UnderscoreTypeNameTransformation(fragmentsByName));
         }
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/OverallQueryTransformer.java]: This `usedHydration` code needs to be worked on later.

File: src/main/java/graphql/nadel/engine/ServiceResultNodesToOverallResult.java [hunk 4]
```diff
@@ -67,8 +67,17 @@ private ListExecutionResultNode mapListExecutionResultNode(ListExecutionResultNo
...
+        FetchedValueAnalysis fetchedValueAnalysis = fetchedAnalysisMapper.mapFetchedValueAnalysis(originalFetchAnalysis, overallSchema, transformationMap);
+
+        objectResultNode = new ObjectExecutionResultNode(fetchedValueAnalysis, objectResultNode.getChildren());
+
+        FieldTransformation fieldTransformation = transformationMap.get(originalField.getSingleField());
+        if (fieldTransformation != null) {
+            objectResultNode = fieldTransformation.unapplyResultNode(objectResultNode);
+        }
+        return objectResultNode;
     }
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/ServiceResultNodesToOverallResult.java]: This is a 2 step mapping - first the FetchedValueAnalysis is changed (say with field renames and __typename being removed) and then the actual result objects are changed (eg remove the __typename) in the response

File: src/main/java/graphql/nadel/engine/transformation/FieldTransformation.java [hunk 5]
```diff
@@ -1,59 +1,19 @@
...
-        //This will be an abstract function
-        throw new UnsupportedOperationException("not ready yet.");
+    default MergedField unapplyMergedField(MergedField mergedField) {
+        return mergedField;
+    }
+
+    default <T extends ExecutionResultNode> T unapplyResultNode(T executionResultNode) {
+        return executionResultNode;
     }
 }
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/transformation/FieldTransformation.java]: We now have new unapplyMethods - one for the field AST and one for the result nodes

File: src/main/java/graphql/nadel/engine/transformation/FieldUtils.java [hunk 6]
```diff
@@ -1,10 +1,84 @@
...
+import java.util.List;
+import java.util.Map;
 
 public final class FieldUtils {
+
     public static String resultKeyForField(Field field) {
         return field.getAlias() != null ? field.getAlias() : field.getName();
     }
 
+    public static boolean hasFieldSubSelectionAtIndex(String fieldName, Field parentField, int desiredIndex) {
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/transformation/FieldUtils.java]: I need to write unit tests for these

File: src/main/java/graphql/nadel/schema/UnderlyingSchemaGenerator.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                    assertNotNull(typeName, "The Nadel result object for interfaces and unions MUST have __typename with a non null value in them in them");
+
+                    GraphQLObjectType objectType = env.getSchema().getObjectType(typeName.toString());
+                    assertNotNull(objectType, "There must be an underlying object type called '%s'", typeName);
+                    return objectType;
+                };
+            }
+        };
+    }
+}
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/schema/UnderlyingSchemaGenerator.java]: This now means that if we send __typename then the standard graphql-java TypeResolvers can be used to get the object type from the schema

File: src/test/groovy/graphql/nadel/NadelE2ETest.groovy [hunk 8]
```diff
@@ -27,7 +28,7 @@ class NadelE2ETest extends Specification {
          }
         '''
 
-    def simpleUnderlyingSchema = TestUtil.schema('''
+    def simpleUnderlyingSchema = typeDefinitions('''
```
Comment (bbakerman) [inline src/test/groovy/graphql/nadel/NadelE2ETest.groovy]: We no longer get a GraphqlSchema back from underlying tests, but rather type defs - hence all fix ups

File: src/main/java/graphql/nadel/ServiceExecutionFactory.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+public interface ServiceExecutionFactory {
+
+    /**
+     * Called to to get a function that can be called to get data for the named service
+     *
+     * @param serviceName the name of the service
+     *
+     * @return a function that can be called to get data from that service
+     */
+    ServiceExecution getDelegatedExecution(String serviceName);
```
Comment (andimarek) [inline src/main/java/graphql/nadel/ServiceExecutionFactory.java]: should rename that too `getSeriveExecution` at one point

File: src/main/java/graphql/nadel/engine/NadelContext.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.UUID;
+
+/**
+ * We sue a wrapper Nadel context object over top of the caller supplied one'
+ * so that we can get access to Nadel specific context properties in
+ * a type safe manner
+ */
+@Internal
+public class NadelContext {
+    final Object callerSuppliedContext;
```
Comment (andimarek) [inline src/main/java/graphql/nadel/engine/NadelContext.java]: I would prefer to identity this as `user....` .

File: src/main/java/graphql/nadel/engine/UnderscoreTypeNameUtils.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+    static Field maybeAddUnderscoreTypeName(NadelContext nadelContext, Field field, GraphQLOutputType fieldType) {
+        if (!Util.isInterfaceOrUnionField(fieldType)) {
+            return field;
+        }
+        String underscoreTypeNameAlias = nadelContext.getUnderscoreTypeNameAlias();
+        assertNotNull(underscoreTypeNameAlias, "We MUST have a generated __typename alias in the request context");
+
+        SelectionSet selectionSet = field.getSelectionSet();
+        Field underscoreTypeNameAliasField = Field.newField("__typename").alias(underscoreTypeNameAlias).build();
```
Comment (andimarek) [inline src/main/java/graphql/nadel/engine/UnderscoreTypeNameUtils.java]: maybe use the constant for __typename

File: src/main/java/graphql/nadel/engine/UnderscoreTypeNameUtils.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+                childNode = maybeRemoveUnderscoreTypeNameImpl(underscoreTypeNameAlias, childNode);
+                newChildren.add(childNode);
+            }
+        }
+        return resultNode.withNewChildren(newChildren);
+    }
+
+    private static boolean isAliasedUnderscoreTpeNameField(String underscoreTypeNameAlias, MergedField mergedField) {
+        List<Field> fields = mergedField.getFields();
+        if (fields.size() == 1) {
```
Comment (andimarek) [inline src/main/java/graphql/nadel/engine/UnderscoreTypeNameUtils.java]: this is a bit tricky implicit assumption which comes from using a UUID, right? maybe worth a comment

File: src/main/java/graphql/nadel/engine/transformation/FieldUtils.java [hunk 13]
```diff
@@ -3,6 +3,14 @@
...
 public final class FieldUtils {
+
+    /**
+     * This returns the aliased result name if a field is alised other its the field name
+     *
+     * @param field the field in play
+     *
+     * @return the result name
+     */
     public static String resultKeyForField(Field field) {
```
Comment (andimarek) [inline src/main/java/graphql/nadel/engine/transformation/FieldUtils.java]: should probably just a method on the `Field` in GraphQL Java

File: src/test/groovy/graphql/nadel/NadelInterfaceAndUnionHandlingTest.groovy [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+
+import static graphql.nadel.Nadel.newNadel
+import static graphql.nadel.NadelExecutionInput.newNadelExecutionInput
+import static graphql.nadel.testutils.TestUtil.schema
+import static graphql.nadel.testutils.TestUtil.serviceFactory
+import static graphql.nadel.testutils.TestUtil.typeDefinitions
+import static graphql.schema.idl.RuntimeWiring.newRuntimeWiring
+import static graphql.schema.idl.TypeRuntimeWiring.newTypeWiring
+
+class NadelInterfaceAndUnionHandlingTest extends Specification {
```
Comment (andimarek) [inline src/test/groovy/graphql/nadel/NadelInterfaceAndUnionHandlingTest.groovy]: I think it should be named ...E2E tests to make it more clear

File: src/main/java/graphql/nadel/engine/NadelExecutionStrategy.java [hunk 15]
```diff
@@ -340,4 +348,13 @@ private ExecutionContext buildServiceExecutionContext(ExecutionContext execution
         );
     }
 
+    private String buildOperationName(Service service) {
```
Comment (andimarek) [inline src/main/java/graphql/nadel/engine/NadelExecutionStrategy.java]: question is a bit: should we also incorporate the original operation name of the overall query if it is exists?

File: src/main/java/graphql/nadel/ServiceExecutionFactory.java (hunk 9 above)
Comment (bbakerman) [reply src/main/java/graphql/nadel/ServiceExecutionFactory.java]: done

File: src/main/java/graphql/nadel/engine/NadelContext.java (hunk 10 above)
Comment (bbakerman) [reply src/main/java/graphql/nadel/engine/NadelContext.java]: k

File: src/main/java/graphql/nadel/engine/UnderscoreTypeNameUtils.java (hunk 12 above)
Comment (bbakerman) [reply src/main/java/graphql/nadel/engine/UnderscoreTypeNameUtils.java]: done

File: src/main/java/graphql/nadel/engine/NadelExecutionStrategy.java (hunk 15 above)
Comment (bbakerman) [reply src/main/java/graphql/nadel/engine/NadelExecutionStrategy.java]: good idea

**Pull request merged**